crates/turborepo-filewatch/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+                                    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
+                                    if event.event.kind == EventKind::Create(CreateKind::Folder) {
+                                        for new_path in &event.event.paths {
+                                            println!("new {}", new_path.display());
+                                            //debouncer.watcher().watch(&new_path, RecursiveMode::Recursive)?;
+                                            watch_recursively(&new_path, debouncer.watcher(), Some((time, &broadcast_sender)))?;
+                                        }
+                                    }
+                                    // we don't care if we fail to send, it just means no one is currently watching
+                                    let _ = broadcast_sender.send(event);
+                                }
+                            },
+                            Err(errors) => {
+                                println!("errors {:?}", errors);
+                                panic!("uh oh")
+                            }
+                        }
+                    }
+                }
+            }
+        });
+        Self {
+            sender,
+            exit_ch
+        }
+    }
+
+    pub fn subscribe(&self) -> broadcast::Receiver<DebouncedEvent> {
+        self.sender.subscribe()
+    }
+}
+
+fn watch_recursively(root: &Path, watcher: &mut RecommendedWatcher, sender: Option<(Instant, &broadcast::Sender<DebouncedEvent>)>) -> Result<(), WatchError> {
+    for dir in WalkDir::new(root).follow_links(false).into_iter() {
+        let dir = dir?;
+        if dir.file_type().is_dir() {
+            watcher.watch(dir.path(), RecursiveMode::NonRecursive)?;
+        }
+        if let Some((instant, sender)) = sender.as_ref() {
+            let create_kind = if dir.file_type().is_dir() {
+                CreateKind::Folder
+            } else {
+                CreateKind::File
+            };
+            let event = DebouncedEvent {
+                event: Event {
+                    paths: vec![dir.path().to_owned()],
+                    kind: EventKind::Create(create_kind),
+                    attrs: EventAttributes::default()
+                },
+                time: *instant
+            };
+            // It's ok if we fail to send, it means we're shutting down
+            let _ = sender.send(event);
+        }
+        println!("ADD {}", dir.path().display());
+    }
+    Ok(())
+}
+
+fn run_watcher(root: &AbsoluteSystemPath, sender: mpsc::Sender<DebounceEventResult>) -> Result<Debouncer<RecommendedWatcher, FileIdMap>, WatchError> {
+    //let (tx, recv) = mpsc::channel();
+    let mut debouncer = new_debouncer(Duration::from_millis(1), None, move |res| {
+        futures::executor::block_on(async {
+            // It's ok if we fail to send, it means we're shutting down
+            let _ = sender.send(res).await;
+        })
```

Comment:
Is [`blocking_send`](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#method.blocking_send) an alternative to this? Also, do we want to warn if the channel fills up and we're unable to instantly send?

Reply:
> `blocking_send` 
yes, I think so. Assuming it works, I like that better.

> channel fills up
My understanding of this scenario is that the sender wouldn't error, but the receiver would. The receiver would lose the oldest messages and get an error. I think in our case we would treat this as "filewatching failed" and invalidate all globs, or whatever the conservative behavior is for whatever filewatching client. https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html#lagging